flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalogUtil.java

```diff
@@ -18,32 +18,179 @@
+	private static final String FLINK_PROPERTY_PREFIX = "flink.";
+
+	// Flink tables should be stored as 'external' tables in Hive metastore
+	private static final Map<String, String> EXTERNAL_TABLE_PROPERTY = new HashMap<String, String>() {{
+		put("EXTERNAL", "TRUE");
+	}};
+
 	private GenericHiveMetastoreCatalogUtil() {
 	}
 
 	// ------ Utils ------
 
 	/**
-	 * Creates a Hive database from CatalogDatabase.
+	 * Creates a Hive database from a CatalogDatabase.
+	 *
+	 * @param databaseName name of the database
+	 * @param catalogDatabase the CatalogDatabase instance
+	 * @return a Hive database
 	 */
-	public static Database createHiveDatabase(String dbName, CatalogDatabase db) {
-		Map<String, String> props = db.getProperties();
+	public static Database createHiveDatabase(String databaseName, CatalogDatabase catalogDatabase) {
 		return new Database(
-			dbName,
-			db.getDescription().isPresent() ? db.getDescription().get() : null,
+			databaseName,
+			catalogDatabase.getDescription().isPresent() ? catalogDatabase.getDescription().get() : null,
 			null,
-			props);
+			catalogDatabase.getProperties());
+	}
+
+	/**
+	 * Creates a Hive table from a CatalogBaseTable.
+	 *
+	 * @param tablePath path of the table
+	 * @param table the CatalogBaseTable instance
+	 * @return a Hive table
+	 */
+	public static Table createHiveTable(ObjectPath tablePath, CatalogBaseTable table) {
+		Map<String, String> properties = new HashMap<>(table.getProperties());
+
+		// Table description
+		if (table.getDescription().isPresent()) {
+			properties.put(HiveTableConfig.TABLE_DESCRITPION, table.getDescription().get());
+		}
+
+		Table hiveTable = new Table();
+		hiveTable.setDbName(tablePath.getDatabaseName());
+		hiveTable.setTableName(tablePath.getObjectName());
+		hiveTable.setCreateTime((int) (System.currentTimeMillis() / 1000));
+
+		// Table properties
+		hiveTable.setParameters(buildFlinkProperties(properties));
+		hiveTable.getParameters().putAll(EXTERNAL_TABLE_PROPERTY);
+
+		// Hive table's StorageDescriptor
+		StorageDescriptor sd = new StorageDescriptor();
+		sd.setSerdeInfo(new SerDeInfo(null, null, new HashMap<>()));
+
+		List<FieldSchema> allColumns = createHiveColumns(table.getSchema());
+
+		// Table columns and partition keys
+		CatalogTable catalogTable = (CatalogTable) table;
+
+		if (catalogTable.isPartitioned()) {
+			int partitionKeySize = catalogTable.getPartitionKeys().size();
+			List<FieldSchema> regularColumns = allColumns.subList(0, allColumns.size() - partitionKeySize);
+			List<FieldSchema> partitionColumns = allColumns.subList(allColumns.size() - partitionKeySize, allColumns.size());
+
+			sd.setCols(regularColumns);
+			hiveTable.setPartitionKeys(partitionColumns);
+		} else {
+			sd.setCols(allColumns);
+			hiveTable.setPartitionKeys(new ArrayList<>());
+		}
+
+		hiveTable.setSd(sd);
+
+		if (table instanceof CatalogTable) {
```

Comment:
Line #107 already assumes table is CatalogTable. With Line #123, it seems unclear what we are trying to achieve. This might need some refinement.

Reply:
good catch. Let me clean it up a bit